include/deal.II/matrix_free/hanging_nodes_internal.h

```diff
@@ -33,6 +33,93 @@ namespace internal
 {
   namespace MatrixFreeFunctions
   {
+    /**
+     * Here is the system for how we store constraint types in a binary mask.
+     * This is not a complete contradiction-free system, i.e., there are
+     * invalid states that we just assume that we never get.
+     *
+     * If the mask is zero, there are no constraints. Then, there are three
+     * different fields with one bit per dimension. The first field determines
+     * the type, or the position of an element along each direction. The
+     * second field determines if there is a constrained face with that
+     * direction as normal. The last field determines if there is a
+     * constrained edge of a given pair of coordinate planes, but where
+     * neither of the corresponding faces are constrained (only valid in 3D).
+     *
+     * The element is placed in the 'first position' along *-axis. These also
+     * determine which face is constrained. For example, in 2D, if
+     * face_x and type are set, then x = 0 is constrained.
+     */
+    enum ConstraintTypes : unsigned short
```

Comment:
What precisely `unsigned short` is depends on the compiler and system. I think you might be better off saying it explicitly that you want a 16-bit type: `std::uint16_t`, see https://en.cppreference.com/w/cpp/types/integer

Strictly speaking, compilers don't *have* to provide these types and one should then rely on `uint_fast16_t`, but I don't think there are compilers that don't provide `uint16_t`.